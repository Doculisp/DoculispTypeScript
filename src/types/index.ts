// Export all types for external consumption
export * from './types.ast';
export * from './types.astDoculisp';
export * from './types.astProject';
export * from './types.containers';
export * from './types.controller';
export * from './types.document';
export * from './types.docWriter';
export * from './types.fileHandler';
export * from './types.filePath';
export * from './types.general';
export * from './types.includeBuilder';
export * from './types.internal';
export * from './types.sringBuilder';
export * from './types.stringWriter';
export * from './types.structure';
export * from './types.textHelpers';
export * from './types.tokens';
export * from './types.trimArray';
export * from './types.variableTable';
export * from './types.version';