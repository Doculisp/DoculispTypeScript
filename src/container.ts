import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};