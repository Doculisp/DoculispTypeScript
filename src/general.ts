import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};