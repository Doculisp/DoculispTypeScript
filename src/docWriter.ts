import { AstBulletStyle, AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "./types.ast";
import { IRegisterable } from "./types.containers";
import { IDocWriter } from "./types.docWriter";
import { IUtil, Result } from "./types.general";

function buildWriter(util: IUtil): IDocWriter {
    function getLabel(style: AstBulletStyle, load: ILoad): string {
        switch (style) {
            case 'bulleted':
            case 'numbered':
            case 'unlabeled':
                return '';
            default:
                return load.sectionLabel;
        }
    }

    function getNumberedBullet(): (label: string, text: string) => string {
        let cnt = 1;
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            const result = `${cnt}. ${lbl}${text}`;
            cnt++;
            return result;
        }

        return bullet;
    }

    function getBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `* ${lbl}${text}`;
        }

        return bullet;
    }

    function getNoBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `${lbl}${text}`;
        }
    }

    function getBulletBuilder(style: AstBulletStyle): (label: string, text: string) => string {
        switch (style) {
            case 'bulleted':
            case 'bulleted-labeled':
                return getBullet();
            case 'numbered':
            case 'numbered-labeled':
                return getNumberedBullet();
            default:
                return getNoBullet();
        }
    }

    function compileToc(astToc: ITableOfContents, loads: ILoad): string {
        if(astToc.bulletStyle === 'no-table') {
            return '';
        }
    }

    function compileContentLocation(loads: ILoad[]): string {
        const parts: string[] = [];

        loads.forEach(load => {
            if(load.document) {
                const section = compileSection(load.document);
                parts.push(section);
            }
        });

        return `${parts.join('\n')}`;
    }

    function compileHeader(astHeader: IHeader): string {
        const padding = '#'.repeat(astHeader.depthCount);
        return `${padding} ${astHeader.text} ${padding}`;
    }

    function compileTitle(astTitle: ITitle): string {
        return astTitle.label;
    }

    function compileWrite(astWrite: IWrite): string {
        return astWrite.value;
    }

    function compilePart(astPart: AstPart) : string {
        astPart.
    }

    function compileSection(astSection: ISectionWriter): string {
        const parts: string[] = [];

        
    }

    function compile(astResult: Result<IAst>): Result<string> {
        if(!astResult.success) {
            return astResult;
        }

        const ast = astResult.value;

        if(ast.section.type === 'ast-empty') {
            return util.ok('');
        }

        throw new Error("Not yet Implemented");
    }

    function write(path: string, astResult: Result<IAst>): Result<string> {
        throw new Error('Not yet Implemented');
    }

    return {
        compile,
        write,
    };
}

const documentWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'docWriter',
    singleton: true,
    dependencies: ['util', 'fileHandler']
};

export {
    documentWriter,
};