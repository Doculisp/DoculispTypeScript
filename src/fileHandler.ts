import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};