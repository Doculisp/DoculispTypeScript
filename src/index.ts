#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
