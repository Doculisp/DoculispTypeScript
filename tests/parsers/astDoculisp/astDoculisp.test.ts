import fs from 'fs';
import path from 'path';
import { Options } from "approvals/lib/Core/Options";
import { configure } from "approvals/lib/config";
import { getVerifier } from "../../tools";
import { container } from "../../../src/container";
import { IDoculisp, IDoculispParser, IEmptyDoculisp } from '../../../src/types/types.astDoculisp'
import { IFail, IProjectLocation, ISuccess, IUtil, Result } from "../../../src/types/types.general";
import { buildLocation, testable } from "../../testHelpers";
import { IAstEmpty, RootAst } from '../../../src/types/types.ast';

describe('astDoculisp', () => {
    let verifyAsJson: (data: any, options?: Options) => void;
    let ok: (successfulValue: any) => ISuccess<any> = undefined as any;
    let fail: (message: string, documentPath: string) => IFail = undefined as any;
    let util: IUtil = undefined as any;
    let toResult: (text: string, projectLocation: IProjectLocation) => Result<IDoculisp | IEmptyDoculisp> = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        toResult = testable.doculisp.resultBuilder(container, environment => {
            util = environment.buildAs<IUtil>('util');
        });
        
        ok = util.ok;
        fail = util.fail;
    });

    describe('basic functionality', () => {
        let parser: IDoculispParser = undefined as any;

        beforeEach(() => {
            util = null as any;
            parser = testable.doculisp.parserBuilder(container, environment => {
                util = environment.buildAs<IUtil>('util');
            });
        });
    
        it('should return an empty doculisp if given empty ast', () => {
            const empty: IAstEmpty = {
                type: 'ast-Empty',
            };
    
            const result = parser.parse(ok(empty));
    
            verifyAsJson(result);
        });

        it('should return failure if given failure', () => {
            const failure = fail('this is a document failure', 'Z:/mybad.dlisp');
    
            const result = parser.parse(failure);
    
            verifyAsJson(result);
        });
    
        it('should parse a value', () => {
            const projectLocation = buildLocation('T:/ext/only.md', 2, 9);

            const ast: RootAst = {
                ast: [
                    {
                        type: 'ast-value',
                        value: 'Some text',
                        location: util.toLocation(projectLocation, 2, 1),
                    }
                ],
                location: projectLocation,
                type: 'RootAst'
            };
    
            const result = parser.parse(ok(ast));
    
            verifyAsJson(result);
        });
    
        it('should parse multiple value ast elements', () => {
            const projectLocation = buildLocation('T:/ext/only.md', 4, 8);
            const ast: RootAst = {
                ast: [
                    {
                        type: 'ast-value',
                        value: 'Intro text',
                        location: util.toLocation(projectLocation, 1, 1),
                    },
                    {
                        type: 'ast-value',
                        value: 'Text after some comment',
                        location: util.toLocation(projectLocation, 5, 1),
                    }
                ],
                type: 'RootAst',
                location: projectLocation,
            };
    
            const result = parser.parse(ok(ast));
    
            verifyAsJson(result);
        });
    });

    describe('lisp', () => {
        it('should simple ast', () => {
            const contents = `<!--
(dl (# My heading))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 2, 1));
    
            verifyAsJson(result);
        });
    
        it('should not parse a bad header', () => {
            const contents = `<!--
(dl (#head My heading))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 3, 2));
    
            verifyAsJson(result);
        });
    
        it('should not parse a header without a parameter', () => {
            const contents = `<!--
(dl (#))
-->`;
            const result = toResult(contents, buildLocation('S:/ome/file.md', 2, 3));
    
            verifyAsJson(result);
        });

        it.skip('should parse a document with all the parts', () => {
            const text = `
(section-meta
    (title Doculisp)
    (include
        (Section ./structure.md)
        (Section ./doculisp.md)
        (Section ./section-meta.md)
        (Section ./content.md)
        (Section ./headings.md)
        (Section ./comment.md)
        (Section ./keywords.md)
    )
)

(## An introduction to Doculisp)

(content (toc numbered-labeled))
`;

            const result = toResult(text, buildLocation('./_main.dlisp', 4, 7));

            verifyAsJson(result);
        });

        describe('section-meta', () => {
            it('should handle all subparts put together out of order', () => {
                const contents = `
(section-meta
(ref-link doculisp_is_)
(include
    (Section ./structure.md)
    (Section ./doculisp.md)
)
(title Doculisp is ✨)
)`;

                const result = toResult(contents, buildLocation('main.dlisp', 1, 4));

                verifyAsJson(result);
            });

            it.skip('should not parse a section-meta that contains a section-meta', () => {
                const content = `
(section-meta
    (section-meta
        (title A Subtitle?)
    )
    (title The Document)
)
`;
                
                const result = toResult(content, buildLocation('A:/malformed/file.dlisp', 1, 7));

                verifyAsJson(result);
            });

            it.skip('should not parse a second section-meta in a file', () => {
                const content = `
(section-meta
    (title My Section)
)

(section-meta
    (title A tale of two sections)
)
`;

                const result = toResult(content, buildLocation('./two/sections.dlisp', 3, 2));

                verifyAsJson(result);
            });

            it.skip('should not parse a section meta with a invalid atom', () => {
                const content = `<!--
(dl
    (section-meta
        (title A Bad atom)
        (bad does stuff)
        (include
                (sample ./good.md)
        )
    )
)
-->

A story of a misbehaving parser.

`

                const result = toResult(content, buildLocation('./_main.md', 1, 1));

                verifyAsJson(result);
            });

            describe('title', () => {
                it('should parse a title', () => {
                    const contents = `
(section-meta
    (title My Cool Document)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 7));
            
                    verifyAsJson(result);
                });
            
                it('should not parse a title without a parameter', () => {
                    const contents = `(section-meta (title))`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 1));
            
                    verifyAsJson(result);
                });

                it('should not parse a title with a sub group', () => {
                    const contents = `(section-meta (title (bad group)))`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 1));
            
                    verifyAsJson(result);

                });

                it('should not parse multiple titles', () => {
                    const contents = '(section-meta (title A Title) (title B Title))';
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 1));

                    verifyAsJson(result);
                });
            });

            describe('ref-link', () => {
                it('should parse the ref-link', () => {
                    const contents = `
(section-meta
    (title My cool title✨)
    (ref-link my_cool_title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
        
                    verifyAsJson(result);
                });

                it('should parse the ref-link if it comes before the title', () => {
                    const contents = `
(section-meta
    (ref-link my_cool_title)
    (title My cool title✨)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
        
                    verifyAsJson(result);
                });

                it('should not parse a ref-link with no parameter', () => {
                    const contents = `
(section-meta
    (ref-link)
    (title My cool title✨)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 10));
                            
                    verifyAsJson(result);
                });

                it('should not parse a ref-link with a sub block', () => {
                    const contents = `
(section-meta
    (title My cool title✨)
    (ref-link (weird block))
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 3, 11));
                    verifyAsJson(result);
                });

                it('should not parse multiple ref-links', () => {
                    const contents = `
(section-meta
    (ref-link my_cool_title)
    (title My cool title✨)
    (ref-link my-cool-title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 1));
                    verifyAsJson(result);
                });

                it('should strip out some special characters from the ref-link', () => {
    //                 const content = `<!--
    // (dl
    //     (section-meta
    //         (title A !@#$%^&*+=\(\){[]}|\\;:'" Story ,./<>? about things)
    //     )
    // )
    // -->`;
                    const content = `<!--
    (dl
        (section-meta
            (title A !@#$%^&*+=;:'" Story ,./<>? about things)
        )
    )
    -->`;
    
                    const result = toResult(content, buildLocation('./_main.md', 1, 1));
                    verifyAsJson(result);
                });
            });

            describe('subtitle', () => {
                it('should parse the subtitle command', () => {
                    const contents = `
(section-meta
    (title My cool title)
    (subtitle This is information)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 4, 4));
        
                    verifyAsJson(result);
                });

                it('should parse the subtitle before title command', () => {
                    const contents = `
(section-meta
    (subtitle This is information)
    (title My cool title)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 2, 7));
        
                    verifyAsJson(result);
                });

                it('should not parse a subtitle without a parameter', () => {
                    const contents = `
(section-meta
    (title My cool title)
    (subtitle)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 10));
        
                    verifyAsJson(result);
                });

                it('should not parse a subtitle with a sub block', () => {
                     const contents = `
(section-meta
    (title My cool title)
    (subtitle (link-ref tom))
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 1, 10));
        
                    verifyAsJson(result);
                })

                it('should not parse multiple subtitles', () => {
                    const contents = `
(section-meta
    (subtitle This is information)
    (title My cool title)
    (subtitle A journey in lisp)
)
`;
                    const result = toResult(contents, buildLocation('main.dlisp', 2, 7));
        
                    verifyAsJson(result);
                });
            });

            describe('include', () => {
                it('should parse include', () => {
                    const contents = `
(section-meta
    (title Doculisp a short description)
    (include
        (section ./doculisp.md)
        (section ./section-meta.md)
    )
)`;

                    const result = toResult(contents, buildLocation('main.dlisp', 1, 4));

                    verifyAsJson(result);
                });

                it('should parse include without section information', () => {
                    const contents = `
(section-meta
    (title Doculisp a short description)
    (include)
)`;

                    const result = toResult(contents, buildLocation('main.dlisp', 1, 4));

                    verifyAsJson(result);
                });

                it('should handle them all put together', () => {
                    const contents = `
(section-meta
    (title Doculisp is ✨)
    (ref-link doculisp_is_)
    (include
        (Section ./structure.md)
        (Section ./doculisp.md)
    )
)`;

                    const result = toResult(contents, buildLocation('main.dlisp', 1, 4));

                    verifyAsJson(result);
                });
            });
        });

        describe.skip('content', () => {
            it.skip('should parse the content location', () => {
                const text = `
(section-meta
    (title Using Content)
    (include
        (Section ./HelloContent.md)
    )
)

(content)
`;

                const result = toResult(text, buildLocation('../main.dlisp', 2, 7));

                verifyAsJson(result);
            });

            it.skip('should not parse the content if it is before the section-meta', () => {
                const text = `
(content)

(section-meta
    (title Using Content)
    (include
        (Section ./HelloContent.md)
    )
)
`;

                const result = toResult(text, buildLocation('../main.dlisp', 4, 2));

                verifyAsJson(result);
            });

            it.skip('should not parse the content location when there are no externals', () => {
                const text = `
(section-meta
    (title Using Content)
)

(content)
`;

                const result = toResult(text, buildLocation('../main.dlisp', 2, 7));

                verifyAsJson(result);
            });

            it.skip('should parse a table of contents', () => {
                const text = `
(section-meta
    (title Sing Me a Song)
    (include
        (Chapter ./sleep.md)
    )
)

(content (toc))
`;

                const result = toResult(text, buildLocation('./itty.dlisp', 2, 1));

                verifyAsJson(result);
            });

            it.each([
                'no-table',
                'unlabeled',
                'labeled',
                'numbered',
                'numbered-labeled',
                'bulleted',
                'bulleted-labeled',
            ])('should parse a table of contents with bullet style of %s', (bulletType: string) => {
                const text = `
(section-meta
    (title Sing Me a Song)
    (include
        (Chapter ./songs.dlisp)
    )
)

(content (toc ${bulletType}))
`;

                const result = toResult(text, buildLocation('./itty.dlisp', 2, 1));

                verifyAsJson(result);
            });

            it.skip('should not parse a content whith a sub command other then toc', () => {
                const text = `
(section-meta
    (title Sing Me a Song)
    (include
        (Chapter ./songs.dlisp)
    )
)

(content (# Incorrect))
`;
                const result = toResult(text, buildLocation('./itty.dlisp', 2, 1));

                verifyAsJson(result);
            });

            it.skip('should not parse a table of contents with unrecognizable bullet style', () => {
                const text = `
(section-meta
    (title Sing Me a Song)
    (include
        (Chapter ./sleep.md)
    )
)

(content (toc unknown))
`;

                const result = toResult(text, buildLocation('./itty.dlisp', 2, 1));

                verifyAsJson(result);
            });
        })
    });

    describe.skip('parse its own documentation', () => {
        function getContents(fileName: string, depth: number, index: number): Result<IDoculisp | IEmptyDoculisp> {
            const filePath = path.join('./documentation/', fileName);
            const location: IProjectLocation = buildLocation(filePath, depth, index);

            const content = fs.readFileSync(filePath, { encoding: 'utf8' });

            return toResult(content, location);
        }

        it.skip('should build ast for structure.md', () => {
            const result = getContents('structure.md', 2, 1);
            verifyAsJson(result);
        });

        it.skip('should build ast for doculisp.md', () => {
            const result = getContents('doculisp.md', 2, 2);
            verifyAsJson(result);
        });

        it.skip('should build ast for section-meta.md', () => {
            const result = getContents('section-meta.md', 2, 3);
            verifyAsJson(result);
        });

        it.skip('should build ast for content.md', () => {
            const result = getContents('content.md', 2, 4);
            verifyAsJson(result);
        });

        it.skip('should build ast for headings.md', () => {
            const result = getContents('headings.md', 2, 5);
            verifyAsJson(result);
        });

        it.skip('should build ast for comment.md', () => {
            const result = getContents('comment.md', 2, 6);
            verifyAsJson(result);
        });

        it.skip('should build ast for keywords.md', () => {
            const result = getContents('keywords.md', 2, 7);
            verifyAsJson(result);
        });

        it.skip('should build ast for contributors.md', () => {
            const result = getContents('contributors.md', 2, 8);
            verifyAsJson(result);
        });

        it.skip('should build ast for _main.md', () => {
            const result = getContents('_main.md', 1, 1);
            verifyAsJson(result);
        });
    });
});