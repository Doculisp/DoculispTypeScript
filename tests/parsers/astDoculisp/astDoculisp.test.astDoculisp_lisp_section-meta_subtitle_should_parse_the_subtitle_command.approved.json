{
  "success": true,
  "value": {
    "projectLocation": {
      "documentDepth": 4,
      "documentIndex": 4,
      "documentPath": "main.dlisp"
    },
    "section": {
      "doculisp": [
        {
          "documentOrder": {
            "char": 6,
            "documentDepth": 4,
            "documentIndex": 4,
            "documentPath": "main.dlisp",
            "line": 3
          },
          "label": "#### My cool title ####",
          "ref_link": "#my-cool-title",
          "subtitle": "###### This is information ######",
          "title": "My cool title",
          "type": "doculisp-title"
        }
      ],
      "documentOrder": {
        "char": 1,
        "documentDepth": 4,
        "documentIndex": 4,
        "documentPath": "main.dlisp",
        "line": 1
      },
      "include": [],
      "type": "doculisp-section"
    },
    "type": "doculisp-root"
  }
}
